documentation/Investigating-FIR-as-an-MLIR-dialect.md

```diff
@@ -0,0 +1,404 @@
+dialect, the MLIR __standard dialect__, and other pre-defined
+dialects, such as the [affine dialect](https://bit.ly/2XYDPNj).
+
+#### Disclaimers
+
+This document is not a tutorial on compilers, Fortran, MLIR, LLVM,
+intermediate representations, abstraction levels, semantics, nor type
+theory.
+
+## Requirements Unchanged
+
+The FIR dialect has the same requirements as spelled out in the
+Fortran IR design document. Specifically, the FIR dialect will
+capture the control flow structure of the Fortran source-level
+program. At the highest level of abstraction, Fortran computations
+will be captured as coarse-grain opaque
+operations corresponding to Fortran expressions
+(with links to `Fortran::evaluate::Expr<T>` objects from the
+front-end), where only the peripheral use-def information is
+exposed. These operations will, of course, necessarily be lowered
+to other various MLIR dialects, where they can be
+optimized.
+The compiler will continue to lower the representation
+of the input with successive transformations
+to MLIR's LLVM IR dialect, the target independent LLVM IR, machine
+IR, and eventually assembly and/or machine code.
+
+## Details
+
+Because MLIR is our adopted framework, our bridge to the FIR
+dialect of MLIR will use the concepts, libraries, coding
+conventions, etc. of the MLIR project. This means some of the class
+names will necessarily be changed from the original FIR design. For
+example, `Program` becomes `Module`, `Procedure` becomes
+`Function`, `BasicBlock` becomes `Block`, and `Statement` becomes
+`Operation`.
+
+Construction of the CFG structure reuses the original FIR pass that
+flattens and linearizes the Fortran parse tree structure prior to
+creating the FIR dialect of MLIR.
+
+`Afforestation` (a class in the original FIR code)
+of the FIR dialect tree structure has to be rewritten
+as the original framework is replaced by that of
+MLIR. Conceptually, the process and objective remains the same.
+
+### FIR Dialect Details
+
+The FIR dialect is a well-defined set of operations, types, etc. that
+captures the executable semantics and state of a correctly specified
+Fortran program. In Fortran, action statements [Fortran 2018 R515] and
+certain constructs specify the behavior of the program. Among these,
+Fortran expressions [Fortran 2018 Clause 10.1] are intrinsic to the
+computation of new values.
+
+#### FIR Operations
+
+Abstract expression operations are higher-level operations that are
+opaque computations in FIR. FIR does not know the exact computations
+involved. However, the framework requires explicit representation of how
+each operation interacts with others in terms of control flow,
+data-flow, and the types of its operands and results. These operations
+must then be lowered to sequences of operations as required by the
+standard MLIR dialect, the LLVM IR dialect, etc.
+
+##### Expression ops
+
+The following lists the FIR abstract expression operations:
+
++    _Apply_Expr_
```

Comment:
Do you mean something like the following?
expr1 = v1 + v2 //v1 at pos1, v2 at pos2
#dict1 (arg1, arg2) = (arg1->pos1; arg2->pos2)
%3 = fir.apply (%1, %2) expr1 dict1